# Predicting sentiment from product reviews


The goal of this first notebook is to explore logistic regression and feature engineering with existing Turi Create functions.

In this notebook you will use product review data from Amazon.com to predict whether the sentiments about a product (from its reviews) are positive or negative.

* Use SFrames to do some feature engineering
* Train a logistic regression model to predict the sentiment of product reviews.
* Inspect the weights (coefficients) of a trained logistic regression model.
* Make a prediction (both class and probability) of sentiment for a new product review.
* Given the logistic regression weights, predictors and ground truth labels, write a function to compute the **accuracy** of the model.
* Inspect the coefficients of the logistic regression model and interpret their meanings.
* Compare multiple logistic regression models.

Let's get started!
    
## Fire up Turi Create

Make sure you have the latest version of Turi Create.

In [1]:
from __future__ import division
import turicreate
import math
import string
import numpy as np

# Data preparation

We will use a dataset consisting of baby product reviews on Amazon.com.

In [2]:
products = turicreate.SFrame('amazon_baby.sframe/')

Now, let us see a preview of what the dataset looks like.

In [3]:
products

Columns:
	name	str
	review	str
	rating	float

Rows: 183531

Data:
+-------------------------------+-------------------------------+--------+
|              name             |             review            | rating |
+-------------------------------+-------------------------------+--------+
|    Planetwise Flannel Wipes   | These flannel wipes are OK... |  3.0   |
|     Planetwise Wipe Pouch     | it came early and was not ... |  5.0   |
| Annas Dream Full Quilt wit... | Very soft and comfortable ... |  5.0   |
| Stop Pacifier Sucking with... | This is a product well wor... |  5.0   |
| Stop Pacifier Sucking with... | All of my kids have cried ... |  5.0   |
| Stop Pacifier Sucking with... | When the Binky Fairy came ... |  5.0   |
| A Tale of Baby's Days with... | Lovely book, it's bound ti... |  4.0   |
| Baby Tracker&reg; - Daily ... | Perfect for new parents. W... |  5.0   |
| Baby Tracker&reg; - Daily ... | A friend of mine pinned th... |  5.0   |
| Baby Tracker&reg; - Daily ... | This has been an easy way ... |  4.0   |
+-------------------------------+-------------------------------+--------+
[183531 rows x 3 columns]
Note: Only the head of the SFrame is printed.
You can use print_rows(num_rows=m, num_columns=n) to print more rows and columns.

## Build the word count vector for each review

Let us explore a specific example of a baby product.


In [4]:
products[269]

{'name': 'The First Years Massaging Action Teether',
 'review': 'A favorite in our house!',
 'rating': 5.0}

Now, we will perform 2 simple data transformations:

1. Remove punctuation using [Python's built-in](https://docs.python.org/2/library/string.html) string functionality.
2. Transform the reviews into word-counts.

**Aside**. In this notebook, we remove all punctuations for the sake of simplicity. A smarter approach to punctuations would preserve phrases such as "I'd", "would've", "hadn't" and so forth. See [this page](https://www.cis.upenn.edu/~treebank/tokenization.html) for an example of smart handling of punctuations.

In [5]:
import string 
def remove_punctuation(text):
    try: # python 2.x
        text = text.translate(None, string.punctuation) 
    except: # python 3.x
        translator = text.maketrans('', '', string.punctuation)
        text = text.translate(translator)
    return text

review_without_punctuation = products['review'].apply(remove_punctuation)
products['word_count'] = turicreate.text_analytics.count_words(review_without_punctuation)

Now, let us explore what the sample example above looks like after these 2 transformations. Here, each entry in the **word_count** column is a dictionary where the key is the word and the value is a count of the number of times the word occurs.

In [6]:
products[269]['word_count']

{'our': 1.0, 'in': 1.0, 'favorite': 1.0, 'house': 1.0, 'a': 1.0}

## Extract sentiments

We will **ignore** all reviews with *rating = 3*, since they tend to have a neutral sentiment.

In [7]:
products = products[products['rating'] != 3]
len(products)

166752

Now, we will assign reviews with a rating of 4 or higher to be *positive* reviews, while the ones with rating of 2 or lower are *negative*. For the sentiment column, we use +1 for the positive class label and -1 for the negative class label.

In [8]:
products['sentiment'] = products['rating'].apply(lambda rating : +1 if rating > 3 else -1)
products

Columns:
	name	str
	review	str
	rating	float
	word_count	dict
	sentiment	int

Rows: 166752

Data:
+-------------------------------+-------------------------------+--------+
|              name             |             review            | rating |
+-------------------------------+-------------------------------+--------+
|     Planetwise Wipe Pouch     | it came early and was not ... |  5.0   |
| Annas Dream Full Quilt wit... | Very soft and comfortable ... |  5.0   |
| Stop Pacifier Sucking with... | This is a product well wor... |  5.0   |
| Stop Pacifier Sucking with... | All of my kids have cried ... |  5.0   |
| Stop Pacifier Sucking with... | When the Binky Fairy came ... |  5.0   |
| A Tale of Baby's Days with... | Lovely book, it's bound ti... |  4.0   |
| Baby Tracker&reg; - Daily ... | Perfect for new parents. W... |  5.0   |
| Baby Tracker&reg; - Daily ... | A friend of mine pinned th... |  5.0   |
| Baby Tracker&reg; - Daily ... | This has been an easy way ... |  4.0   |
| Baby Tracker&reg; - Daily ... | I love this journal and ou... |  4.0   |
+-------------------------------+-------------------------------+--------+
+-------------------------------+-----------+
|           word_count          | sentiment |
+-------------------------------+-----------+
| {'recommend': 1.0, 'disapp... |     1     |
| {'quilt': 1.0, 'this': 1.0... |     1     |
| {'tool': 1.0, 'clever': 1.... |     1     |
| {'rock': 1.0, 'headachesth... |     1     |
| {'thumb': 1.0, 'or': 1.0, ... |     1     |
| {'2995': 1.0, 'for': 1.0, ... |     1     |
| {'right': 1.0, 'because': ... |     1     |
| {'like': 1.0, 'and': 1.0, ... |     1     |
| {'in': 1.0, 'pages': 1.0, ... |     1     |
| {'tracker': 1.0, 'now': 1.... |     1     |
+-------------------------------+-----------+
[166752 rows x 5 columns]
Note: Only the head of the SFrame is printed.
You can use print_rows(num_rows=m, num_columns=n) to print more rows and columns.

Now, we can see that the dataset contains an extra column called **sentiment** which is either positive (+1) or negative (-1).

## Split data into training and test sets

Let's perform a train/test split with 80% of the data in the training set and 20% of the data in the test set. We use `seed=1` so that everyone gets the same result.

In [9]:
train_data, test_data = products.random_split(.8, seed=1)
print(len(train_data))
print(len(test_data))

133416
33336


# Train a sentiment classifier with logistic regression

We will now use logistic regression to create a sentiment classifier on the training data. This model will use the column **word_count** as a feature and the column **sentiment** as the target. We will use `validation_set=None` to obtain same results as everyone else.

**Note:** This line may take 1-2 minutes.

In [10]:
sentiment_model = turicreate.logistic_classifier.create(train_data,
                                                        target = 'sentiment',
                                                        features=['word_count'],
                                                        validation_set=None)

Logistic regression:

--------------------------------------------------------

Number of examples          : 133416

Number of classes           : 2

Number of feature columns   : 1

Number of unpacked features : 121712

Number of coefficients      : 121713

Starting L-BFGS

--------------------------------------------------------

+-----------+----------+-----------+--------------+-------------------+

| Iteration | Passes   | Step size | Elapsed Time | Training Accuracy |

+-----------+----------+-----------+--------------+-------------------+

| 0         | 4        | 0.250000  | 1.791197     | 0.840754          |

| 1         | 9        | 3.250000  | 2.825198     | 0.941514          |

| 2         | 11       | 2.778177  | 3.310720     | 0.942638          |

| 3         | 12       | 2.778177  | 3.619302     | 0.967822          |

| 4         | 13       | 2.778177  | 3.945284     | 0.976495          |

| 5         | 14       | 2.778177  | 4.253790     | 0.976495          |

+-----------+----------+-----------+--------------+-------------------+

In [11]:
sentiment_model

Class                          : LogisticClassifier

Schema
------
Number of coefficients         : 121713
Number of examples             : 133416
Number of classes              : 2
Number of feature columns      : 1
Number of unpacked features    : 121712

Hyperparameters
---------------
L1 penalty                     : 0.0
L2 penalty                     : 0.01

Training Summary
----------------
Solver                         : lbfgs
Solver iterations              : 5
Solver status                  : TERMINATED: Terminated due to numerical difficulties.
Training time (sec)            : 2.6047

Settings
--------
Log-likelihood                 : inf

Highest Positive Coefficients
-----------------------------
word_count[offsi]              : 21.7657
word_count[kidsgood]           : 21.6818
word_count[conclusions]        : 21.6818
word_count[easycheap]          : 21.6818
word_count[torsional]          : 21.6818

Lowest Negative Coefficients
----------------------------
word_count[wahwah]

**Aside**. You may get a warning to the effect of "Terminated due to numerical difficulties --- this model may not be ideal". It means that the quality metric (to be covered in Module 3) failed to improve in the last iteration of the run. The difficulty arises as the sentiment model puts too much weight on extremely rare words. A way to rectify this is to apply regularization, to be covered in Module 4. Regularization lessens the effect of extremely rare words. For the purpose of this assignment, however, please proceed with the model above.

Now that we have fitted the model, we can extract the weights (coefficients) as an SFrame as follows:

In [12]:
weights = sentiment_model.coefficients
weights.column_names()

['name', 'index', 'class', 'value', 'stderr']

In [13]:
weights

Columns:
	name	str
	index	str
	class	int
	value	float
	stderr	float

Rows: 121713

Data:
+-------------+--------------+-------+------------------------+--------+
|     name    |    index     | class |         value          | stderr |
+-------------+--------------+-------+------------------------+--------+
| (intercept) |     None     |   1   |   0.7797362814629109   |  None  |
|  word_count |  recommend   |   1   |   0.3052457737572151   |  None  |
|  word_count | disappointed |   1   |  -2.4267924944818877   |  None  |
|  word_count |     wise     |   1   |  -0.18170817410434445  |  None  |
|  word_count |     love     |   1   |   0.8405057320622243   |  None  |
|  word_count |      it      |   1   |  0.010479446044282479  |  None  |
|  word_count |    planet    |   1   |  -0.3846729911246308   |  None  |
|  word_count |     and      |   1   |  0.03631437278666848   |  None  |
|  word_count |     bags     |   1   |   0.1305759809889736   |  None  |
|  word_count |    wipes     |   1   | -0.0020751803185189623 |  None  |
+-------------+--------------+-------+------------------------+--------+
[121713 rows x 5 columns]
Note: Only the head of the SFrame is printed.
You can use print_rows(num_rows=m, num_columns=n) to print more rows and columns.

There are a total of `121713` coefficients in the model. Recall from the lecture that positive weights $w_j$ correspond to weights that cause positive sentiment, while negative weights correspond to negative sentiment. 

Fill in the following block of code to calculate how many *weights* are positive ( >= 0). (**Hint**: The `'value'` column in SFrame *weights* must be positive ( >= 0)).

In [14]:
num_positive_weights = (weights['value'] >= 0).sum()
num_negative_weights = (weights['value'] < 0).sum()
print("Number of positive weights: %s " % num_positive_weights)
print("Number of negative weights: %s " % num_negative_weights)

Number of positive weights: 91073 
Number of negative weights: 30640 


**Quiz Question:** How many weights are >= 0?

Ans: 91073

## Making predictions with logistic regression

Now that a model is trained, we can make predictions on the **test data**. In this section, we will explore this in the context of 3 examples in the test dataset.  We refer to this set of 3 examples as the **sample_test_data**.

In [16]:
sample_test_data = test_data[10:13]
print(sample_test_data['rating'])
sample_test_data

[5.0, 2.0, 1.0]


Columns:
	name	str
	review	str
	rating	float
	word_count	dict
	sentiment	int

Rows: 3

Data:
+-------------------------------+-------------------------------+--------+
|              name             |             review            | rating |
+-------------------------------+-------------------------------+--------+
|   Our Baby Girl Memory Book   | Absolutely love it and all... |  5.0   |
| Wall Decor Removable Decal... | Would not purchase again o... |  2.0   |
| New Style Trailing Cherry ... | Was so excited to get this... |  1.0   |
+-------------------------------+-------------------------------+--------+
+-------------------------------+-----------+
|           word_count          | sentiment |
+-------------------------------+-----------+
| {'again': 1.0, 'book': 1.0... |     1     |
| {'peeling': 1.0, '5': 1.0,... |     -1    |
| {'on': 1.0, 'waste': 1.0, ... |     -1    |
+-------------------------------+-----------+
[3 rows x 5 columns]

Let's dig deeper into the first row of the **sample_test_data**. Here's the full review:

In [17]:
sample_test_data[0]['review']

'Absolutely love it and all of the Scripture in it.  I purchased the Baby Boy version for my grandson when he was born and my daughter-in-law was thrilled to receive the same book again.'

That review seems pretty positive.

Now, let's see what the next row of the **sample_test_data** looks like. As we could guess from the sentiment (-1), the review is quite negative.

In [18]:
sample_test_data[1]['review']

'Would not purchase again or recommend. The decals were thick almost plastic like and were coming off the wall as I was applying them! The would NOT stick! Literally stayed stuck for about 5 minutes then started peeling off.'

We will now make a **class** prediction for the **sample_test_data**. The `sentiment_model` should predict **+1** if the sentiment is positive and **-1** if the sentiment is negative. Recall from the lecture that the **score** (sometimes called **margin**) for the logistic regression model  is defined as:

$$
\mbox{score}_i = \mathbf{w}^T h(\mathbf{x}_i)
$$ 

where $h(\mathbf{x}_i)$ represents the features for example $i$.  We will write some code to obtain the **scores** using Turi Create. For each row, the **score** (or margin) is a number in the range **[-inf, inf]**.

In [19]:
scores = sentiment_model.predict(sample_test_data, output_type='margin')
print(scores)

[4.788907309210134, -3.0007822224448573, -8.18850136076355]


### Predicting sentiment

These scores can be used to make class predictions as follows:

$$
\hat{y} = 
\left\{
\begin{array}{ll}
      +1 & \mathbf{w}^T h(\mathbf{x}_i) > 0 \\
      -1 & \mathbf{w}^T h(\mathbf{x}_i) \leq 0 \\
\end{array} 
\right.
$$

Using scores, write code to calculate $\hat{y}$, the class predictions:

In [20]:
predictions = scores.apply(lambda scores: +1 if scores> 0 else -1)
predictions

dtype: int
Rows: 3
[1, -1, -1]

Run the following code to verify that the class predictions obtained by your calculations are the same as that obtained from Turi Create.

In [21]:
print("Class predictions according to Turi Create:")
print(sentiment_model.predict(sample_test_data))

Class predictions according to Turi Create:
[1, -1, -1]


**Checkpoint**: Make sure your class predictions match with the one obtained from Turi Create.

### Probability predictions

Recall from the lectures that we can also calculate the probability predictions from the scores using:
$$
P(y_i = +1 | \mathbf{x}_i,\mathbf{w}) = \frac{1}{1 + \exp(-\mathbf{w}^T h(\mathbf{x}_i))}.
$$

Using the variable **scores** calculated previously, write code to calculate the probability that a sentiment is positive using the above formula. For each row, the probabilities should be a number in the range **[0, 1]**.

In [22]:
probability_predictions = scores.apply(lambda scores: (1/(1+math.exp(-scores))))
probability_predictions

dtype: float
Rows: 3
[0.9917471313286568, 0.047390547487918794, 0.0002777527712170442]

**Checkpoint**: Make sure your probability predictions match the ones obtained from Turi Create.

In [23]:
print("Class predictions according to Turi Create:")
print(sentiment_model.predict(sample_test_data, output_type='probability'))

Class predictions according to Turi Create:
[0.9917471313286567, 0.04739054748791879, 0.0002777527712170442]


** Quiz Question:** Of the three data points in **sample_test_data**, which one (first, second, or third) has the **lowest probability** of being classified as a positive review?

Ans: 3

# Find the most positive (and negative) review

We now turn to examining the full test dataset, **test_data**, and use Turi Create to form predictions on all of the test data points for faster performance.

Using the `sentiment_model`, find the 20 reviews in the entire **test_data** with the **highest probability** of being classified as a **positive review**. We refer to these as the "most positive reviews."

To calculate these top-20 reviews, use the following steps:
1.  Make probability predictions on **test_data** using the `sentiment_model`. (**Hint:** When you call `.predict` to make predictions on the test data, use option `output_type='probability'` to output the probability rather than just the most likely class.)
2.  Sort the data according to those predictions and pick the top 20. (**Hint:** You can use the `.topk` method on an SFrame to find the top k rows sorted according to the value of a specified column.)

In [26]:
test_data['probability']= sentiment_model.predict(test_data, output_type= 'probability').sort(ascending = False)
# predictions.sort(ascending= False)

In [68]:
# test_data.topk('probability',20,False)['name']
test_data.sort('probability', ascending=False)['review'].unique()[0:20]

dtype: str
Rows: 20
['Unlike some of the other reviewers, we haven\'t had any problems with this swing at all; in fact, it has worked very well. I love that it plugs in, and I actually chose it for the "drab" colors, as ours sits in the living room and not the nursery. It\'s a bit classier looking than its garish primary-colored cousins. One reviewer mentioned that the tray gets in the way, but it snaps off and on very easily.  The mobile swings back out of the way, as well, so it\'s very easy to get baby in and out.There are two things I would probably change about this swing, but neither is a major concern for us. First, there isn\'t a noticeable difference among the six speeds, at least not on ours. The one speed it has, though, is perfect for calming our baby, so I really can\'t complain. Secondly, the music really grates on my nerves. The sound of it is cheap and cheesey, and even on the lower volume setting it seems much too loud for being as close to the baby\'s head as it is. B

In [ ]:
print()

In [50]:
print(type(test_data[3]['name']))

<class 'str'>


**Quiz Question**: Which of the following products are represented in the 20 most positive reviews? [multiple choice]


Now, let us repeat this exercise to find the "most negative reviews." Use the prediction probabilities to find the  20 reviews in the **test_data** with the **lowest probability** of being classified as a **positive review**. Repeat the same steps above but make sure you **sort in the opposite order**.

In [70]:
test_data.topk('probability',20,True)['name']

dtype: str
Rows: 20
['Best 2 Pack Baby Car Shade for Kids - Window Shade Baby - Cheapest Baby Sun Shade with Excellent Quality - Great Protection Car Window Shade for Travel, Trips, or Just for Shade - Better Than Britax and Brica - Pop Open Window Shade Cling That Folds to the Size of Your Hand - Twice As Strong As Competitors - Lifetime Money Back Guarantee', 'Squeasy Snacker 6oz Silicone Reusable Food Pouch (Turquoise)', 'Airline Seat Belt Extender - The Best Extension for All Airlines - Airplane Seatbelt Buckle Is Compatible with American, Alaska, Continental, Delta, Frontier, Jetblue, Us Airways, United Airlines and Many More - Great for Men, Women, and Kids - This Is a Novelty - Lifetime Money Back Guarantee!', 'Maxboost iPhone 5S/5 Case - Protective Snap-on Hard Case Slim Rugged Cover [Not compatible to Apple iPhone 6 Air 5c 4s 4 3gs, Screen Protector / Cable is not included] - Ultra Slim Profile Slimmer than coventional otterbox/lifeproof/kate Spade/speck/juicy couture/griffin/

In [40]:
test_data.sort('probability', ascending=True)['name'].unique()[0:20]

dtype: str
Rows: 20
['CM Star Candy Heart Crossbody Shoulder Bag Satchel Purse Handbag Clutch', 'Medela Breastmilk Storage and Feeding Set', 'Summer Infant Dual View Digital Color Video Monitor', 'Zippy Fun Bandana Bibs for Babies and Toddlers (Retro Cool for Boys) (Pack of 4)', 'SoHo Nap Mat , Dinosaurs', 'Summer Infant  Multi Use Extra Tall Walk-Thru Gate, White', 'My NapPak Play Ball Napmat', '2 in 1 Head Support By Gold Bug - Pink/Gray', 'Tiny Love Super Mat', 'Trend Lab Set of 2 Frame Set , Sweet Safari Blue Print', 'Wimmer Ferguson Wimmer Soothing Silicone Teether', 'ERGObaby Teething Pads', 'Summer Infant Swaddleme 100% Cotton Knit - 4 Pack, Jungle Fun (Large)', "Oryer Cars Kids Backpack School Bag Children Boy's Xmas Gift", 'Cocooi Merino Baby Sleep Sack, Mint, For Newborn Babies', '8oz Bare Air-free baby bottles. With Syringe-like Air-plug&reg;, feeds baby Air-free milk to help prevent gas &amp; colic. Perfe-latch&reg; Nipple promotes breastfeeding latch, stretches &amp; dispe

**Quiz Question**: Which of the following products are represented in the 20 most negative reviews?  [multiple choice]

In [39]:
test_data.topk('probability',20,False)['name']

dtype: str
Rows: 20
['Our Baby Girl Memory Book', "Milestones, Kids' Step Stone Kit", 'Neurosmith - Music Blocks with Mozart Music Cartridge', 'Modern House Safari Animal Family removable Vinyl Mural Art Wall Sticker Decal', 'Modern House Giraffe Children Height Measurement Growth Chart Vinyl Mural Art Wall Sticker Decal', 'Edu-Petit Happy Jungle Day Finger Puppet Book Developmental Toy', 'Large Jungle Animals (Panda, Lion, Elephant, Gorilla) - 3D Wall Murals for Baby Nursery and Kids Rooms - Made in USA', 'Newborn Baby Tracker&reg; - Round the Clock Childcare Journal, Log Book', 'New Style Trailing Cherry Blossom Tree Decal Removable Vinyl Kids Wall Decal Wall Sticker Peel and Stick', 'Wall Decor Removable Decal Sticker - Colorful Butterflies', 'Baby Tracker&reg; - Daily Childcare Journal, Schedule Log', "Baby's First Year Undated Wall Calendar with Stickers to Mark &quot;Firsts&quot; - Pink", "SoftPlay Peek-A-Boo Where's Elmo A Children's Book", 'SoftPlay Giggle Jiggle Funbook, Happy

## Compute accuracy of the classifier

We will now evaluate the accuracy of the trained classifier. Recall that the accuracy is given by


$$
\mbox{accuracy} = \frac{\mbox{# correctly classified examples}}{\mbox{# total examples}}
$$

This can be computed as follows:

* **Step 1:** Use the trained model to compute class predictions (**Hint:** Use the `predict` method)
* **Step 2:** Count the number of data points when the predicted class labels match the ground truth labels (called `true_labels` below).
* **Step 3:** Divide the total number of correct predictions by the total number of data points in the dataset.

Complete the function below to compute the classification accuracy:

In [71]:
def get_classification_accuracy(model, data, true_labels):
    # First get the predictions
    ## YOUR CODE HERE
    predictions = model.predict(data)
    
    # Compute the number of correctly classified examples
    ## YOUR CODE HERE
    num_correct_predictions = (predictions==true_labels).sum()

    # Then compute accuracy by dividing num_correct by total number of examples
    ## YOUR CODE HERE
    accuracy = num_correct_predictions/len(data)
    
    return accuracy

Now, let's compute the classification accuracy of the **sentiment_model** on the **test_data**.

In [72]:
get_classification_accuracy(sentiment_model, test_data, test_data['sentiment'])


0.9221862251019919

**Quiz Question**: What is the accuracy of the **sentiment_model** on the **test_data**? Round your answer to 2 decimal places (e.g. 0.76).

Ans: 0.92

**Quiz Question**: Does a higher accuracy value on the **training_data** always imply that the classifier is better?

Ans: No

## Learn another classifier with fewer words

There were a lot of words in the model we trained above. We will now train a simpler logistic regression model using only a subset of words that occur in the reviews. For this assignment, we selected a 20 words to work with. These are:

In [74]:
significant_words = ['love', 'great', 'easy', 'old', 'little', 'perfect', 'loves', 
      'well', 'able', 'car', 'broke', 'less', 'even', 'waste', 'disappointed', 
      'work', 'product', 'money', 'would', 'return']

In [75]:
len(significant_words)

20

For each review, we will use the **word_count** column and trim out all words that are **not** in the **significant_words** list above. We will use the [SArray dictionary trim by keys functionality]( https://dato.com/products/create/docs/generated/graphlab.SArray.dict_trim_by_keys.html). Note that we are performing this on both the training and test set.

In [76]:
train_data['word_count_subset'] = train_data['word_count'].dict_trim_by_keys(significant_words, exclude=False)
test_data['word_count_subset'] = test_data['word_count'].dict_trim_by_keys(significant_words, exclude=False)

Let's see what the first example of the dataset looks like:

In [77]:
train_data[0]['review']

'it came early and was not disappointed. i love planet wise bags and now my wipe holder. it keps my osocozy wipes moist and does not leak. highly recommend it.'

The **word_count** column had been working with before looks like the following:

In [78]:
print(train_data[0]['word_count'])

{'recommend': 1.0, 'disappointed': 1.0, 'wise': 1.0, 'love': 1.0, 'it': 3.0, 'planet': 1.0, 'and': 3.0, 'bags': 1.0, 'wipes': 1.0, 'highly': 1.0, 'not': 2.0, 'early': 1.0, 'came': 1.0, 'i': 1.0, 'does': 1.0, 'my': 2.0, 'was': 1.0, 'now': 1.0, 'wipe': 1.0, 'holder': 1.0, 'leak': 1.0, 'keps': 1.0, 'osocozy': 1.0, 'moist': 1.0}


Since we are only working with a subset of these words, the column **word_count_subset** is a subset of the above dictionary. In this example, only 2 `significant words` are present in this review.

In [79]:
print(train_data[0]['word_count_subset'])

{'disappointed': 1.0, 'love': 1.0}


## Train a logistic regression model on a subset of data

We will now build a classifier with **word_count_subset** as the feature and **sentiment** as the target. 

In [80]:
simple_model = turicreate.logistic_classifier.create(train_data,
                                                     target = 'sentiment',
                                                     features=['word_count_subset'],
                                                     validation_set=None)
simple_model

Logistic regression:

--------------------------------------------------------

Number of examples          : 133416

Number of classes           : 2

Number of feature columns   : 1

Number of unpacked features : 20

Number of coefficients      : 21

Starting Newton Method

--------------------------------------------------------

+-----------+----------+--------------+-------------------+

| Iteration | Passes   | Elapsed Time | Training Accuracy |

+-----------+----------+--------------+-------------------+

| 1         | 2        | 0.310392     | 0.862917          |

| 2         | 3        | 0.457604     | 0.865713          |

| 3         | 4        | 0.570635     | 0.866478          |

| 4         | 5        | 0.680756     | 0.866748          |

| 5         | 6        | 0.792482     | 0.866815          |

| 6         | 7        | 0.901878     | 0.866815          |

+-----------+----------+--------------+-------------------+

SUCCESS: Optimal solution found.

Class                          : LogisticClassifier

Schema
------
Number of coefficients         : 21
Number of examples             : 133416
Number of classes              : 2
Number of feature columns      : 1
Number of unpacked features    : 20

Hyperparameters
---------------
L1 penalty                     : 0.0
L2 penalty                     : 0.01

Training Summary
----------------
Solver                         : newton
Solver iterations              : 6
Solver status                  : SUCCESS: Optimal solution found.
Training time (sec)            : 0.9222

Settings
--------
Log-likelihood                 : 44323.7254

Highest Positive Coefficients
-----------------------------
word_count_subset[loves]       : 1.6773
word_count_subset[perfect]     : 1.5145
word_count_subset[love]        : 1.3654
(intercept)                    : 1.2995
word_count_subset[easy]        : 1.1937

Lowest Negative Coefficients
----------------------------
word_count_subset[disappointed] : -2.3551
wo

We can compute the classification accuracy using the `get_classification_accuracy` function you implemented earlier.

In [81]:
get_classification_accuracy(simple_model, test_data, test_data['sentiment'])

0.8693004559635229

Now, we will inspect the weights (coefficients) of the **simple_model**:

In [82]:
simple_model.coefficients

Columns:
	name	str
	index	str
	class	int
	value	float
	stderr	float

Rows: 21

Data:
+-------------------+--------------+-------+---------------------+
|        name       |    index     | class |        value        |
+-------------------+--------------+-------+---------------------+
|    (intercept)    |     None     |   1   |  1.2995449552026996 |
| word_count_subset | disappointed |   1   |  -2.355092500610724 |
| word_count_subset |     love     |   1   |  1.3654354936790394 |
| word_count_subset |     well     |   1   |  0.5042567463979308 |
| word_count_subset |   product    |   1   | -0.3205554929955735 |
| word_count_subset |    loves     |   1   |  1.6772714555592931 |
| word_count_subset |    little    |   1   |   0.52062863602502  |
| word_count_subset |     work     |   1   |  -0.621700012425313 |
| word_count_subset |     easy     |   1   |  1.1936618983284664 |
| word_count_subset |    great     |   1   |  0.9446912694798442 |
+-------------------+--------------+-------+---------------------+
+----------------------+
|        stderr        |
+----------------------+
| 0.012088854133054213 |
| 0.05041498885569795  |
| 0.030354629510905326 |
| 0.021381300630990176 |
| 0.01543113213620174  |
| 0.04823282753835029  |
| 0.02146914756649054  |
| 0.023033059794584882 |
| 0.02928886920202954  |
| 0.020950992659050264 |
+----------------------+
[21 rows x 5 columns]
Note: Only the head of the SFrame is printed.
You can use print_rows(num_rows=m, num_columns=n) to print more rows and columns.

Let's sort the coefficients (in descending order) by the **value** to obtain the coefficients with the most positive effect on the sentiment.

In [83]:
simple_model.coefficients.sort('value', ascending=False).print_rows(num_rows=21)

+-------------------+--------------+-------+----------------------+
|        name       |    index     | class |        value         |
+-------------------+--------------+-------+----------------------+
| word_count_subset |    loves     |   1   |  1.6772714555592931  |
| word_count_subset |   perfect    |   1   |  1.5144862670271362  |
| word_count_subset |     love     |   1   |  1.3654354936790394  |
|    (intercept)    |     None     |   1   |  1.2995449552026996  |
| word_count_subset |     easy     |   1   |  1.1936618983284664  |
| word_count_subset |    great     |   1   |  0.9446912694798442  |
| word_count_subset |    little    |   1   |   0.52062863602502   |
| word_count_subset |     well     |   1   |  0.5042567463979308  |
| word_count_subset |     able     |   1   |  0.1914383022947517  |
| word_count_subset |     old      |   1   | 0.08539618866781735  |
| word_count_subset |     car      |   1   | 0.05883499006802087  |
| word_count_subset |     less     |   1   | -0.

In [85]:
simple_pos_weights = simple_model.coefficients['index'][simple_model.coefficients['value']>0]
len(simple_pos_weights)

11

**Quiz Question**: Consider the coefficients of **simple_model**. There should be 21 of them, an intercept term + one for each word in **significant_words**. How many of the 20 coefficients (corresponding to the 20 **significant_words** and *excluding the intercept term*) are positive for the `simple_model`?

Ans: 10

In [86]:
sentiment_pos_weights = sentiment_model.coefficients['index'][sentiment_model.coefficients['value']>0]
len(sentiment_pos_weights)

91073

In [87]:
set1 = set(simple_pos_weights)
set2 = set(sentiment_pos_weights)
set1.issubset(set2)

True

**Quiz Question**: Are the positive words in the **simple_model** (let us call them `positive_significant_words`) also positive words in the **sentiment_model**?

Ans: True

# Comparing models

We will now compare the accuracy of the **sentiment_model** and the **simple_model** using the `get_classification_accuracy` method you implemented above.

First, compute the classification accuracy of the **sentiment_model** on the **train_data**:

In [88]:
get_classification_accuracy(sentiment_model, train_data, train_data['sentiment'])

0.976494573364514

Now, compute the classification accuracy of the **simple_model** on the **train_data**:

In [89]:
get_classification_accuracy(simple_model, train_data, train_data['sentiment'])

0.8668150746537147

**Quiz Question**: Which model (**sentiment_model** or **simple_model**) has higher accuracy on the TRAINING set?

Ans: sentiment model

Now, we will repeat this exercise on the **test_data**. Start by computing the classification accuracy of the **sentiment_model** on the **test_data**:

In [90]:
sent_accuracy = get_classification_accuracy(sentiment_model, test_data, test_data['sentiment'])
sent_accuracy

0.9221862251019919

Next, we will compute the classification accuracy of the **simple_model** on the **test_data**:

In [91]:
simp_accuracy = get_classification_accuracy(simple_model, test_data, test_data['sentiment'])
simp_accuracy

0.8693004559635229

**Quiz Question**: Which model (**sentiment_model** or **simple_model**) has higher accuracy on the TEST set?

Ans: sentiment model

## Baseline: Majority class prediction

It is quite common to use the **majority class classifier** as the a baseline (or reference) model for comparison with your classifier model. The majority classifier model predicts the majority class for all data points. At the very least, you should healthily beat the majority class classifier, otherwise, the model is (usually) pointless.

What is the majority class in the **train_data**?

In [92]:
num_positive  = (train_data['sentiment'] == +1).sum()
num_negative = (train_data['sentiment'] == -1).sum()
print(num_positive)
print(num_negative)

112164
21252


In [93]:
maj_accuracy = (test_data['sentiment'] == +1).sum()/len(test_data)
maj_accuracy

0.8427825773938085

Now compute the accuracy of the majority class classifier on **test_data**.

**Quiz Question**: Enter the accuracy of the majority class classifier model on the **test_data**. Round your answer to two decimal places (e.g. 0.76).

Ans: 0.84

In [94]:
sent_accuracy> maj_accuracy

True

**Quiz Question**: Is the **sentiment_model** definitely better than the majority class classifier (the baseline)?

Ans: True